### PR TITLE
Add file-evening to hadoop pre-process

### PR DESCRIPTION
**Pull request by jenniferdai**
* Currently, pre-process only allows users to set max.num.records. Now, what may happen is that we may have very skewed segment sizes if there is a tiny remainder in the segments. 
* This PR still guarantees max.num.records is the upper bound of record sin a segment but fixes the possible skew by calculating approximately-even segment record numbers prior to mandating the number of records that go into a reducer.

### CODE REVIEW

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java [hunk 1]
```diff
@@ -306,7 +309,15 @@ public void run()
       // so that all the rows can be spread evenly.
       addHashCodeField(fieldSet);
     }
-    setMaxNumRecordsConfigIfSpecified(job);
+    String enablePartitioningString = job.getConfiguration().get(JobConfigConstants.ENABLE_PARTITIONING);
```
Comment (Jackie-Jiang) [inline pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java]: boolean enablePartitioning = job.getConfiguration().getBoolean(JobConfigConstants.ENABLE_PARTITIONING, false);

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java [hunk 2]
```diff
@@ -412,17 +423,40 @@ private void fetchResizingConfig() {
     }
   }
 
-  private void setMaxNumRecordsConfigIfSpecified(Job job) {
+  // TODO: Handle case where #records exceed long max
```
Comment (Jackie-Jiang) [inline pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java]: Exceed long max? We don't need to consider Long overflow

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java [hunk 3]
```diff
@@ -446,6 +480,21 @@ private Schema getSchema(Path inputPathDir)
     return avroSchema;
   }
 
+  private long countAvroRecords(List<Path> inputPaths) throws IOException {
+    long count = 0;
+
+    for (Path path : inputPaths) {
+      DataFileStream<GenericRecord> dataFileStream = getAvroReader(path);
```
Comment (Jackie-Jiang) [inline pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java]: We need to scan all the files to get the record count? That will be super expensive

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java (hunk 2 above)
Comment (jenniferdai) [reply pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java]: maybe.. for refresh use cases i coluld see it

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java (hunk 3 above)
Comment (jenniferdai) [reply pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java]: this seems to be the best approach from my research, do you have other suggestions?

**Comment by jenniferdai** [comment]
others have benchmarked - MapReduce
When I use MapReduce job to iterate through all the records it completes in less than 2 minutes for about million records

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java]: That is more than the number of molecules on earth lol...

**Comment by Jackie-Jiang** [comment]
> others have benchmarked - MapReduce
> When I use MapReduce job to iterate through all the records it completes in less than 2 minutes for about million records

So for 1 billion records, we need to wait for 2000 minutes (more than 1 day) just for the count?

**Pull request closed without merging**